File: implementations/lua/oshash.lua
#!/usr/bin/env lua
-- OpenSubtitles Hash (OSHash) - Lua implementation.
-- Works on both big and little endian architectures.

local function compute_hash(fileName)
    local fil = io.open(fileName, "rb")
    if not fil then
        io.stderr:write("Error opening file: " .. fileName .. "\n")
        os.exit(1)
    end

    local lo, hi = 0, 0

    -- Hash first 64KB (8192 x 8 bytes)
    for i = 1, 8192 do
        local a, b, c, d = fil:read(4):byte(1, 4)
        lo = lo + a + b * 256 + c * 65536 + d * 16777216
        a, b, c, d = fil:read(4):byte(1, 4)
        hi = hi + a + b * 256 + c * 65536 + d * 16777216
        while lo >= 4294967296 do
            lo = lo - 4294967296
            hi = hi + 1
        end
        while hi >= 4294967296 do
            hi = hi - 4294967296
        end
    end

    -- Hash last 64KB
    local size = fil:seek("end", -65536) + 65536
    for i = 1, 8192 do
        local a, b, c, d = fil:read(4):byte(1, 4)
        lo = lo + a + b * 256 + c * 65536 + d * 16777216
        a, b, c, d = fil:read(4):byte(1, 4)
        hi = hi + a + b * 256 + c * 65536 + d * 16777216
        while lo >= 4294967296 do
            lo = lo - 4294967296
            hi = hi + 1
        end
        while hi >= 4294967296 do
            hi = hi - 4294967296
        end
    end

    -- Add file size. Split into 32-bit halves and carry in O(1): a subtraction
    -- loop here would run size/2^32 times (billions for a multi-EB file), and
    -- lo + size could also overflow Lua's 64-bit integer.
    hi = hi + (size // 4294967296)
    lo = lo + (size % 4294967296)
    hi = (hi + (lo // 4294967296)) % 4294967296
    lo = lo % 4294967296

    fil:close()
    return string.format("%08x%08x", hi, lo), size
end

if not arg[1] then
    io.stderr:write("Usage: lua oshash.lua <file>\n")
    os.exit(1)
end

local hash = compute_hash(arg[1])
print(hash)

File: implementations/r/oshash.R
#!/usr/bin/env Rscript
# OpenSubtitles Hash (OSHash) - R implementation.
# Uses split 32-bit arithmetic to handle 64-bit values.

CHUNK_SIZE <- 65536L

hex4 <- function(val) {
  # Convert 0-65535 to 4-char hex
  h <- format(as.hexmode(as.integer(val)))
  paste0(paste(rep("0", 4 - nchar(h)), collapse = ""), h)
}

to_hex8 <- function(val) {
  val <- floor(val) %% (2^32)
  paste0(hex4(floor(val / 65536)), hex4(val %% 65536))
}

compute_hash <- function(filepath) {
  fsize <- file.info(filepath)$size
  if (fsize < CHUNK_SIZE * 2) stop(paste("File too small:", fsize, "bytes"))

  con <- file(filepath, "rb")
  on.exit(close(con))

  # Split into hi/lo 32-bit halves
  lo <- fsize %% (2^32)
  hi <- floor(fsize / (2^32))

  add_to_hash <- function(bytes8) {
    v_lo <- as.numeric(bytes8[1]) + as.numeric(bytes8[2]) * 256 +
            as.numeric(bytes8[3]) * 65536 + as.numeric(bytes8[4]) * 16777216
    v_hi <- as.numeric(bytes8[5]) + as.numeric(bytes8[6]) * 256 +
            as.numeric(bytes8[7]) * 65536 + as.numeric(bytes8[8]) * 16777216

    lo <<- lo + v_lo
    hi <<- hi + v_hi

    if (lo >= 2^32) {
      hi <<- hi + floor(lo / 2^32)
      lo <<- lo %% 2^32
    }
    if (hi >= 2^32) {
      hi <<- hi %% 2^32
    }
  }

  # Hash first 64KB
  for (i in seq_len(8192)) {
    bytes8 <- readBin(con, "raw", n = 8)
    add_to_hash(bytes8)
  }

  # Hash last 64KB
  seek(con, max(0, fsize - CHUNK_SIZE))
  for (i in seq_len(8192)) {
    bytes8 <- readBin(con, "raw", n = 8)
    add_to_hash(bytes8)
  }

  paste0(to_hex8(hi), to_hex8(lo))
}

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) {
  cat("Usage: Rscript oshash.R <file>\n", file = stderr())
  quit(status = 1)
}

cat(compute_hash(args[1]), "\n")
